File: shared/src/commonMain/kotlin/com/mdgroup/nasawallpapers/data/mappers/WallpaperMapper.kt
package com.mdgroup.nasawallpapers.data.mappers

import com.mdgroup.nasawallpapers.data.responses.WallpaperResponse
import com.mdgroup.nasawallpapers.domain.models.WallpaperModel
import com.mdgroup.nasawallpapers.sqldelight.WallpaperEntity

object WallpaperMapper {

    fun responseToModel(response: WallpaperResponse): WallpaperModel {
        return WallpaperModel(
            response.copyright,
            response.date,
            response.explanation,
            response.hdurl ?: "",
            response.mediaType,
            response.serviceVersion,
            response.title ?: "",
            response.url
        )
    }

    fun entityToModel(entity: WallpaperEntity): WallpaperModel {
        return WallpaperModel(
            entity.copyright,
            entity.date,
            entity.explanation,
            entity.hdurl ?: "",
            entity.mediaType,
            entity.serviceVersion,
            entity.title ?: "",
            entity.url,
            entity.uri
        )
    }
}

File: iosApp/iosApp/domain/WallpaperMapper.swift
//
//  WallpaperMapper.swift
//  iosApp
//
//  Created by Pavel Kravchenko on 18.02.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

extension WallpaperIdentifiable {
    
    static func fromModel(model: WallpaperModel) throws -> WallpaperIdentifiable {
        let documentDirectory = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
        
        var path: URL?
        if let uri = model.uri {
            if #available(iOS 16.0, *) {
                path = documentDirectory.appending(path: uri)
            } else {
                path = documentDirectory.appendingPathComponent(uri, conformingTo: .image)
            }
        }
        
        return WallpaperIdentifiable(
            copyright: model.copyright,
            date: model.date,
            explanation: model.explanation,
            hdurl: model.hdurl,
            mediaType: model.mediaType,
            serviceVersion: model.serviceVersion,
            title: model.title,
            url: model.url,
            uri: path?.path ?? nil
        )
    }
}
